# Notebook Jupyter adapté pour le nettoyage et la transformation des jeux de données 

In [2]:
import json
import csv
import pandas as pd

### - Jeu de données ghostbikes (accident mortel)

Ce jeu de données est composé, pour Grenoble, de deux ghostbikes que nous allons ajouter à nos données d'accidentologie en tant qu'accident mortel. En effet, un ghostbike représente une stèle sous forme d'un vélo blanc en hommage aux victimes d'accidents de la route à vélo. Il est important de noter que ce jeu de données provient d'un fichier json récupéré grâce à une requête overpass turbo provenant d'Open Street Map. Nous allons donc commencer par transformer les données en isolant les variables d'intérêts et nous allons finir par créer un fichier final nettoyé.

##### 1. Transformation

In [8]:
# Fonction pour charger un fichier .geojson
def load_geojson(file_path):
    with open(file_path, "r", encoding="utf-8") as file:
        return json.load(file)

# Fonction pour nettoyer les données
def clean_data(data):
    cleaned_data = []
    for feature in data.get("features", []):
        properties = feature.get("properties", {})
        geometry = feature.get("geometry", {})
        coordinates = geometry.get("coordinates", [])
        if len(coordinates) == 2:
            cleaned_data.append({
                "id": properties.get("@id"),
                "lat": coordinates[1],
                "lon": coordinates[0],
                "start_date": properties.get("start_date")
            })
    return cleaned_data

# Fonction pour écrire les données nettoyées dans un fichier .csv
def write_to_csv(cleaned_data, output_file):
    with open(output_file, mode="w", newline="", encoding="utf-8") as file:
        writer = csv.DictWriter(file, fieldnames=["id", "lat", "lon", "start_date"])
        writer.writeheader()
        writer.writerows(cleaned_data)

# Chemin d'entrée et de sortie
input_file = "export.geojson"
output_file = "ghostbikes.csv"

# Chargement, nettoyage et exportation
data = load_geojson(input_file)
cleaned_data = clean_data(data)
write_to_csv(cleaned_data, output_file)

Initialement, le jeu de données est composé de plus de variables. En revanche, dans notre cas, seules la date et les coordonnées géographiques nous seront utiles pour la suite. Nous avons donc ciblé et récupéré les identifiants, la latitude, la longitude et la date de l'accident mortel.

##### 2. Finalisation

In [11]:
data_ghostbike = pd.read_csv("ghostbikes.csv", sep=",") # Importation du jeu de données
df_ghostbike = pd.DataFrame(data_ghostbike) # Transformation en dataframe
df_ghostbike["id"] = df_ghostbike["id"].str.replace("node/", "") # Suppression de "node/" dans la colonne "id"
df_ghostbike["statut_accident"] = "Mortel" # Ajout de la colonne 'statut_accident' avec la valeur "Mortel"
print(df_ghostbike)

            id        lat       lon  start_date statut_accident
0   9611135117  45.180153  5.716451  2022-03-26          Mortel
1  11002892805  45.177253  5.718046  2023-06-24          Mortel


Voici les données des deux accidents qui seront ajoutés au jeu de données de base des accidents en tant qu'accident mortel.

### - Jeu de données accident de vélo

Nous allons donc commencer à explorer ce jeu de données en effectuant des statistiques descriptives pour identifier les traitements nécessaires pour la suite. Puis, nous ajouterons des variables qui seront intéréssantes à utiliser dans le cadre d'un tableau de bord. Pour finir, nous remplacerons les données numériques d'échelle par les variables textuelles réelles à l'aide du descriptif des données. A savoir que ce jeu de données provient de jeu de données publique des accidents de vélo, disponible sur le site data.gouv.

##### 1. Filtrage pour la commune de Grenoble (Code INSEE : 38185)

Afin de réduire les données en vue d'optimiser le nettoyage, nous allons commencer par filtrer les données selon le code INSEE de la commune de Grenoble (38185).

In [14]:
data_accident_velo = pd.read_csv("accidentsVelo.csv", sep=",") # Importation du jeu de données
df_accident_velo = pd.DataFrame(data_accident_velo)

# Filtrer les données selon le code INSEE de la commune de grenoble (38185)
filtered_df = df_accident_velo[df_accident_velo["com"].isin(["38185", "74010", "33063", "67482"])]

C:\Users\alexm\AppData\Local\Temp\ipykernel_5044\1794056342.py:1: DtypeWarning: Columns (8,9,20,21,30) have mixed types. Specify dtype option on import or set low_memory=False.
  data_accident_velo = pd.read_csv("accidentsVelo.csv", sep=",") # Importation du jeu de données


##### 2. Statistiques descriptives

In [5]:
filtered_df[filtered_df["com"] == "38185"].info()

<class 'pandas.core.frame.DataFrame'>
Index: 337 entries, 136 to 67403
Data columns (total 39 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Num_Acc          337 non-null    int64  
 1   date             337 non-null    object 
 2   an               337 non-null    int64  
 3   mois             337 non-null    object 
 4   jour             337 non-null    object 
 5   hrmn             337 non-null    object 
 6   dep              337 non-null    object 
 7   com              337 non-null    object 
 8   lat              337 non-null    object 
 9   long             337 non-null    object 
 10  agg              337 non-null    int64  
 11  int              337 non-null    int64  
 12  col              337 non-null    float64
 13  lum              337 non-null    int64  
 14  atm              337 non-null    float64
 15  catr             337 non-null    int64  
 16  circ             337 non-null    float64
 17  nbv              

En analysant le jeu de données, nous pouvons voir que nous avons 337 enregistrements, soit 337 accidents de vélo sur la commune de Grenoble. En ajoutant les accidents mortels des ghostbikes, nous aurons donc 339 accidents. Ce jeu de données est composé de 39 variables ayant toutes des caractéristiques particulières. Quelques variables possèdent des valeurs manquantes. Aucun traitement ne sera réalisé sur ces données manquantes car ce sont des variables catégorielles et les données seront conservé malgré les données manquantes car, au vue du peu de données que nous avons, chaque observation est très importante pour évaluer des accidents.

##### 3. Standardisation

In [15]:
# Remplacement des "," en "." pour rendre les données traitable numériquement
filtered_df["lat"] = filtered_df["lat"].astype(str).str.replace(",", ".")
filtered_df["long"] = filtered_df["long"].astype(str).str.replace(",", ".")

# Convertir les colonnes en type float pour une manipulation correcte
filtered_df["lat"] = filtered_df["lat"].astype(float)
filtered_df["long"] = filtered_df["long"].astype(float)

C:\Users\alexm\AppData\Local\Temp\ipykernel_5044\821980889.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df["lat"] = filtered_df["lat"].astype(str).str.replace(",", ".")
C:\Users\alexm\AppData\Local\Temp\ipykernel_5044\821980889.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df["long"] = filtered_df["long"].astype(str).str.replace(",", ".")
C:\Users\alexm\AppData\Local\Temp\ipykernel_5044\821980889.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice f

Les données numériques du jeu de données sont au format français. C'est-à-dire que le séparateur des décimales sont des points. En revanche, pour les rendre numériquement traitable par la machine, il est nécéssaire de remplacer les "," par des ".".

##### 4. Création de colonnes pour mesurer la qualité des données

La création de variables pour mesurer la qualité des données est une étape importante car nous avons basé nos analyses sur deux axes, l'axe général des accidents et l'axe géographique. En revanche, nous avons remarqué que beaucoup de données étaient composées de dates et de coordonnées géographiques erronnées. Il est donc nécéssaire, pour filtrer les données pertinentes ou non, d'avoir accès à des variables de qualité. Nous allons donc créer un statut pour les heures et les cordonnées "latitude" et "longitude".

###### Analyse des heures

In [16]:
# Fonction pour vérifier si l'heure est cohérente
def verifier_heure(heure):
    try:
        # Séparer l'heure et les minutes
        h, m = map(int, heure.split(':'))
        # Vérifier si l'heure et les minutes sont dans les plages valides
        return "Cohérente" if 0 <= h <= 23 and 0 <= m <= 59 else "Non cohérente"
    except:
        return "Non cohérente"

# Appliquer la fonction sur la colonne hrmn pour créer une nouvelle colonne statut_heure
filtered_df["statut_heure"] = filtered_df["hrmn"].astype(str).apply(verifier_heure)
print(filtered_df)

            Num_Acc        date    an      mois      jour   hrmn dep    com  \
113    200500003153  2005-01-06  2005   janvier     jeudi  17:15  33  33063   
114    200500003162  2005-01-11  2005   janvier     mardi  14:55  33  33063   
117    200500003179  2005-01-13  2005   janvier     jeudi  17:00  33  33063   
118    200500003185  2005-01-15  2005   janvier    samedi  22:20  33  33063   
121    200500003200  2005-01-18  2005   janvier     mardi  15:20  33  33063   
...             ...         ...   ...       ...       ...    ...  ..    ...   
71512  202000044495  2020-07-28  2020   juillet     mardi  10:55  74  74010   
71531  202000044878  2020-02-04  2020   février     mardi  19:00  74  74010   
71899  202100006031  2021-11-28  2021  novembre  dimanche  00:10  74  74010   
72661  202100022011  2021-08-28  2021      août    samedi  14:00  74  74010   
73206  202100031116  2021-07-02  2021   juillet  vendredi  16:10  33  33063   

             lat      long  ...  equipement  obs  o

C:\Users\alexm\AppData\Local\Temp\ipykernel_5044\281969821.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df["statut_heure"] = filtered_df["hrmn"].astype(str).apply(verifier_heure)


In [17]:
print(filtered_df["statut_heure"].value_counts()) # Evaluation de la qualité des données horaires

statut_heure
Cohérente        2250
Non cohérente     741
Name: count, dtype: int64


Nous avons 96 accidents possédant des heures incohérentes 

###### Analyse des coordonnées

In [18]:
# Fonction pour vérifier la cohérence des coordonnées
def verifier_coord(lat, lon):
    if lat == 0 or lon == 0:
        return "Non cohérentes"
    if lat < 45 or lat > 46:
        return "Non cohérentes"
    if lon < 5 or lon > 6:
        return "Non cohérentes"
    return "Cohérentes"

# Appliquer la fonction ligne par ligne pour créer une nouvelle colonne
filtered_df["statut_coord"] = filtered_df.apply(lambda row: verifier_coord(row["lat"], row["long"]), axis=1)
print(filtered_df)

            Num_Acc        date    an      mois      jour   hrmn dep    com  \
113    200500003153  2005-01-06  2005   janvier     jeudi  17:15  33  33063   
114    200500003162  2005-01-11  2005   janvier     mardi  14:55  33  33063   
117    200500003179  2005-01-13  2005   janvier     jeudi  17:00  33  33063   
118    200500003185  2005-01-15  2005   janvier    samedi  22:20  33  33063   
121    200500003200  2005-01-18  2005   janvier     mardi  15:20  33  33063   
...             ...         ...   ...       ...       ...    ...  ..    ...   
71512  202000044495  2020-07-28  2020   juillet     mardi  10:55  74  74010   
71531  202000044878  2020-02-04  2020   février     mardi  19:00  74  74010   
71899  202100006031  2021-11-28  2021  novembre  dimanche  00:10  74  74010   
72661  202100022011  2021-08-28  2021      août    samedi  14:00  74  74010   
73206  202100031116  2021-07-02  2021   juillet  vendredi  16:10  33  33063   

             lat      long  ...  obs  obsm  choc  m

C:\Users\alexm\AppData\Local\Temp\ipykernel_5044\4172450745.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df["statut_coord"] = filtered_df.apply(lambda row: verifier_coord(row["lat"], row["long"]), axis=1)


In [19]:
print(filtered_df["statut_coord"].value_counts()) # Evaluation de la qualité des données des coordonnées géographiques

statut_coord
Non cohérentes    2927
Cohérentes          64
Name: count, dtype: int64


Nous avons 273 accidents possédant des coordonnées géographiques incohérentes. L'axe géographique sera donc constitué de moins de données.

##### 5. Utilisation des métadonnées pour remplacer les codes numériques par leur signification

In [20]:
# Pour agg
filtered_df["agg"] = filtered_df["agg"].replace({
    1: "Hors agglomération",
    2: "En agglomération"
})

# Pour int
filtered_df["int"] = filtered_df["int"].replace({
    0: "Non renseigné",
    1: "Hors intersection",
    2: "Intersection en X",
    3: "Intersection en T",
    4: "Intersection en Y",
    5: "Intersection à plus de 4 branches",
    6: "Giratoire",
    7: "Place",
    8: "Passage à niveau",
    9: "Autre intersection"
})

# Pour lum
filtered_df["lum"] = filtered_df["lum"].replace({
    1: "Plein jour",
    2: "Crépuscule ou aube",
    3: "Nuit sans éclairage public",
    4: "Nuit avec éclairage public non allumé",
    5: "Nuit avec éclairage public allumé"
})

# Pour atm
filtered_df["atm"] = filtered_df["atm"].replace({
    1: "Normale",
    2: "Pluie légère",
    3: "Pluie forte",
    4: "Neige - grêle",
    5: "Brouillard - fumée",
    6: "Vent fort - tempête",
    7: "Temps éblouissant",
    8: "Temps couvert",
    9: "Autre",
    -1: "Non renseigné"
})

# Pour circ
filtered_df["circ"] = filtered_df["circ"].replace({
    1: "A sens unique",
    2: "Bidirectionnelle",
    3: "A chaussées séparées",
    4: "Avec voies d'affectation variable",
    -1: "Non renseigné"
})

# Pour prof
filtered_df["prof"] = filtered_df["prof"].replace({
    1: "Plat",
    2: "Pente",
    3: "Sommet de côte",
    4: "Bas de côte",
    -1: "Non renseigné"
})

# Pour surf
filtered_df["surf"] = filtered_df["surf"].replace({
    1: "Normale",
    2: "Mouillée",
    3: "Flaques",
    4: "Inondée",
    5: "Enneigée",
    6: "Boue",
    7: "Verglacée",
    8: "Corps gras - huile",
    9: "Autre",
    -1: "Non renseigné"
})

# Pour situ
filtered_df["situ"] = filtered_df["situ"].replace({
    0: "Aucun",
    1: "Sur chaussée",
    2: "Sur bande d'arrêt d'urgence",
    3: "Sur accotement",
    4: "Sur trottoir",
    5: "Sur piste cyclable",
    6: "Sur autre voie spéciale",
    8: "Autres",
    -1: "Non renseigné"
})

# Pour grav
filtered_df["grav"] = filtered_df["grav"].replace({
    1: "Indemne",
    2: "Tué",
    3: "Blessé hospitalisé",
    4: "Blessé léger"
})

# Pour sexe
filtered_df["sexe"] = filtered_df["sexe"].replace({
    1: "Masculin",
    2: "Féminin"
})

# Pour trajet
filtered_df["trajet"] = filtered_df["trajet"].replace({
    0: "Non renseigné",
    1: "Domicile - travail",
    2: "Domicile - école",
    3: "Courses - achats",
    4: "Utilisation professionnelle",
    5: "Promenade - loisirs",
    9: "Autre",
    -1: "Non renseigné"
})

C:\Users\alexm\AppData\Local\Temp\ipykernel_5044\2167096452.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df["agg"] = filtered_df["agg"].replace({
C:\Users\alexm\AppData\Local\Temp\ipykernel_5044\2167096452.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df["int"] = filtered_df["int"].replace({
C:\Users\alexm\AppData\Local\Temp\ipykernel_5044\2167096452.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,co

Nous avons remplacé les données d'échelles des variables identifiées comme les plus importantes par leur valeur réelle à l'aide du fichier "accidents-velos-schema.json" qui représente les métadonnées. En faisant cela, les résultats du tableau de bord seront plus lisibles et compréhensibles par les lecteurs/utilisateurs.

In [21]:
# Ajout des données des ghostbikes à l'aide d'informations trouvées sur les accidents via les médias
data_ghostbikes_ajout = [
    [
        "202200000001", "2021-03-25", "2021", "mars", "jeudi", "", "38", "38185", "45.180153", "5.716451", 
        "En agglomération", "Non renseigné", "-1", "Plein jour", "Non renseigné", "0", "Non renseigné", "-1", "Non renseigné", "-1", "", "", 
        "Non renseigné", "-1", "Non renseigné", "Tué", "Féminin", "38", "Domicile - travail", "3", "", "-1", "-1", "-1", "-1", "", "0", "-1", "0", 
        "Non cohérente", "Cohérentes"
    ],
    [
        "202300000001", "2022-06-13", "2022", "juin", "lundi", "", "38", "38185", "45.177253", "5.718046", 
        "En agglomération", "Non renseigné", "-1", "Plein jour", "Non renseigné", "0", "Non renseigné", "-1", "Non renseigné", "-1", "", "", 
        "Non renseigné", "-1", "Non renseigné", "Tué", "Masculin", "14", "Domicile - école", "3", "", "-1", "-1", "-1", "-1", "", "0", "-1", "0", 
        "Non cohérente", "Cohérentes"
    ]
]

column_names = filtered_df.columns.tolist() # Identification des colonnes du dataframe
data_df_ajout = pd.DataFrame(data_ghostbikes_ajout, columns=column_names) # Conversion des données en dataframe

df_filtered = pd.concat([filtered_df, data_df_ajout], ignore_index=True) # Ajout des données au DataFrame existant

In [22]:
df_filtered.to_csv("accidentvelo_nettoye.csv") # Sauvergarde du fichier final au format .csv